### PR TITLE
Asynchronous lifecycle `change_state`

### PR DESCRIPTION
**Pull request by tgroechel**
This is a draft PR before opening a PR against rclcpp/ros2:rolling. 

The below text is going in the issue. The PR text will link to the issue. Links are removed from the below text as to not reference issues until the final issue/PR are up.

---
## Issue Text

# Asynchronous lifecycle `change_state`

## Feature request
Allow for user transition callbacks within lifecycle nodes (e.g., `on_configure`) to defer work. This would allow for:
1. nodes/entities with external transition dependencies to defer work until those dependencies are met (e.g., requests to external service(s) is/are ready/complete)
2. attempting recovery mid-transition instead of needing to fully shutdown + fully bring the node back to a desired state

## Feature description
Lifecycle transition callbacks currently must be synchronous. With the limitation of calling service within a service PR773 and the advent of async services PR1709, we would like to propose adding the ability to defer work within a user transition callback (e.g., `on_configure`). The `change_state` process would ideally be deferable + cancelleable. This would allow for both long running functions (e.g., a long load in `on_configure`) as well as free up the executor to process new events.

[Demo Repo](https://github.com/tgroechel/lifecycle_prac/tree/defer_cancel_user_transtion_callbacks)

<details>

<summary>Expand for in-line example code</summary>

```cpp
void on_configure_async(
    const rclcpp_lifecycle::State &,
    std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl) {
  // Cancel monitoring
  transition_cancel_monitoring_timer_ = create_wall_timer(
      std::chrono::milliseconds{100}, [this, change_state_hdl]() {
        if (!change_state_hdl->is_executing()) {
          transition_cancel_monitoring_timer_.reset();
          return;
        } else if (change_state_hdl->is_canceling()) { /*handle cancel*/
          size_t num_pruned_req = client_->prune_pending_requests();
          RCLCPP_INFO(this->get_logger(), "Handle cancel: pruned %ld request",
                      num_pruned_req);
          change_state_hdl->handle_canceled(true);
          transition_cancel_monitoring_timer_.reset();
        }
      });

  // Callback for future response of getting a parameter
  auto response_received_callback =
      [logger = this->get_logger(), change_state_hdl](
          rclcpp::Client<rcl_interfaces::srv::GetParameters>::SharedFuture
              future) {
        if (change_state_hdl->is_executing()) {
          auto request_response_pair = future.get();
          RCLCPP_INFO(logger, "Received parameter response: %s",
                      request_response_pair->values[0].string_value.c_str());
          change_state_hdl->send_callback_resp(
              rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
                  CallbackReturn::SUCCESS);
        }
      };

  // Sending the request and attaching the callback
  auto request =
      std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
  request->names.push_back("param1");
  RCLCPP_INFO(this->get_logger(), "Sending async param request");
  client_->async_send_request(request, std::move(response_received_callback));
} // Off executor

void on_activate_async(
  const rclcpp_lifecycle::State & state,
  std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl)
{
  LifecycleNode::on_activate(
    state);   // activates managed entities (i.e., lifecycle_publishers)
  // create a thread to do some work passing the change_state_hdl to the
  // thread
  std::thread t(&LifecycleTalker::defer_on_activate_work, this,
    change_state_hdl);
  t.detach();
}

void defer_on_activate_work(
  std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl)
{
  int sleep_time = 3;
  RCLCPP_INFO(
    this->get_logger(),
    "on_activate() {async} is called, sleeping is separate thread "
    "for %d seconds.",
    sleep_time);
  std::this_thread::sleep_for(std::chrono::seconds{sleep_time});
  RCLCPP_INFO(
    this->get_logger(),
    "on_activate() done sleeping, returning success");
  change_state_hdl->send_callback_resp(
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
    CallbackReturn::SUCCESS);
}
```

</details>

https://github.com/tgroechel/lifecycle_prac/assets/15292506/55600d76-9cea-40b4-8423-76d81e7e7fbf

The "ideal" would be to turn the `ChangeState.srv` into an `Action` ([docs](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Actions/Understanding-ROS2-Actions.html)). This would give the robust capability of accept/reject a change request, publish its status, and have the ability to cancel mid transition with user attempt to handle the cancel. However, turning `change_state` into an action would break all backward compatibility (i.e., all lifecycle nodes and clients).

**We propose an initial stage, backward compatible solution here**: PRLINK

Exact details can be found within that PR but to summarize:
- To make current lifecycle dev as well as transitioning to an `Action` in the future easier, we re-architect `LifecycleNodeInterfaceImpl` for a cleaner separation of concerns similar to [model-view-controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller):
  - `LifecycleNodeInterfaceImpl`: ("`controller`") owner of:
    - `StateManager`: (`model:lifecycle_state`) responsible for lifecycle `state_machine`
    - `StateServicesManager`: ("`view + controller`") public interface of services
    - `EntitiesManager`: (`model:managed_entities`)responsible for entity state
- Created a `ChangeStateHandler` that acts very similar to a `GoalHandle` within a ROS 2 Action. The example above should give better context for this.
  - `shared_ptr` given to the user transition callback for the user to `send_callback_resp(CallbackReturn)` whenever the transition completes
  - has an equivalent `is_cancelled()` so the user can attempt to handle a cancel mid-transition
  - when a `transition_cancel` succeeds, the state follows the respective `CallbackReturn::FAILURE` path (e.g., `CONFIGURING` -> `handled_cancel(true)` -> `UNCONFIGURED`). 

<details>

<summary>Reference for `Callback::FAILURE` paths of lifecycle nodes</summary>

image

</details>

We hope this issue can serve as a discussion around the ability to create asynchronous state changes, whether that be through the above suggested approach and/or a long term plan to turn them into `Actions`.

---

## PR Text

# Async Lifecycle `change_state` Implementation

This PR addresses ISSUELINK to enable async/deferrable user transition callbacks (e.g., `on_configure`) within lifecycle nodes. This is a proposed first-stage solution:
- maintains backward compatibility (the core reason for not directly converting to a ROS 2 Action)
- attempts to structure the code to allow for possible future transition to using a [ROS 2 Action](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Actions/Understanding-ROS2-Actions.html) for `ChangeState`

Dependency: RCL_INTERFACESPRLINK

Demo: DEMOLINK

## Separation of Concerns
The `LifecycleNodeInterfaceImpl` was:
1. managing `ManagedEntities`
3. managing underlying `rcl::state_machine`
4. managing services

<details>

<summary>Separation of concerns / new impl architecture</summary>

soc_lifecycle

</details>

This PR splits these into 3 separate classes as shown in the figure above. This follows more closely to a [model-view-controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) design:

`LifecycleNodeInterfaceImpl`: ("`controller`") owner of:
- `EntitiesManager`: (`model:managed_entities`)responsible for entity state
- `StateManager`: (`model:lifecycle_state`) responsible for lifecycle `state_machine`
- `StateServicesManager`: ("`view + controller`") public interface of services

## ChangeStateHandler
An async transition follows the same flow as before but now passes a `ChangeState` handler that:
- allows response deferral
- allows for handling a cancel request
This is very similar to a `GoalHandler` within a ROS 2 Action. However, we want to maintain backward compatibility. 

### Async transition

<details>

<summary>Control flow of an async transition</summary>

image-20230531-133549

</details>

The above image outlines the process. Only 1 request can be processed at a given time, all other requests are rejected. The `ChangeStateHandler` allows for the user to send a response whenever they are done with the transition (accomplished by passing a newly created `shared_ptr<ChangeStateHandler>` to the user).

### Cancelling a transition

<details>

<summary>Reference for `Callback::FAILURE` paths of lifecycle nodes</summary>

image

</details>

Given that the transition is async, it would be ideal if it could also have the potential to be cancelled. The goal would to attempt to recover the node into a valid primary state. With this:
- only the user can handle a cancel safely so the `ChangeStateHandler` is "marked" as cancelled and waits until the user marks the cancel as completed
- if handled successfully, the state follows a `CallbackReturn::Failure` path. This is desired to
  - recover, mostly, to the last primary state (see reference diagram above)
  - maintain `state_machine` validity/backward compatibility
  - reuse the same code as `change_state`

<details>

<summary>Control flow of an async transition</summary>

cancel_no_timeout

</details>

The example flow of a cancelled transition. Note this is all possible even on a `SingleThreadedExecutor` as the timer acts as a monitor for the transition. Note the rejection and failure criteria / reasons are passed back to the client within the `CancelTransition.srv`.

---

### Additional features that may be added

Features 1 & 2 would be somewhat/theoretically trivial to add given the implementation

1. Ability to directly bind a `handle_cancel` function that is called upon cancel.
2. In the case we are:
    1. in transition
    2. receive a cancel request but we fail to cancel
    3. transition completes
    4. Instead of going to the next state, we **instead treat the mid-transition cancel as a `CallbackReturn::ERROR` and process it as such**
3. [not as trivial] Create a container that can send out a set of requests, creating future callbacks for each. The container would have the ability to them know when all requests have been completed/send the `CallbackReturn` response. The container could also help with handling a cancel request by pruning all waiting request.

### CODE REVIEW

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  virtual bool is_canceling() const = 0;
+
+  // Check to see if the response has been sent
+  /**
+   * @return true if response has not been sent
+   */
+  virtual bool is_executing() const = 0;
+
+  virtual ~ChangeStateHandler() {}
```
Comment (alsora) [inline rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: nitpick

```suggestion
  virtual ~ChangeStateHandler() = default;
```

File: rclcpp_lifecycle/src/change_state_handler_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+: state_manager_hdl_(state_manager_hdl)
+{
+}
+
+bool
+ChangeStateHandlerImpl::send_callback_resp(
+  node_interfaces::LifecycleNodeInterface::CallbackReturn cb_return_code)
+{
+  if (!is_canceling() && is_executing()) {
+    if (auto state_manager_hdl = state_manager_hdl_.lock()) {
```
Comment (alsora) [inline rclcpp_lifecycle/src/change_state_handler_impl.cpp]: minor: I prefer to avoid assignments inside if statements unless there's a major reason for doing that.

this will compile to the same exact code and it's easier to read
```
auto state_manager_hdl = state_manager_hdl_.lock();
if (state_manager_hdl) {
```

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  virtual bool send_callback_resp(
+    node_interfaces::LifecycleNodeInterface::CallbackReturn cb_return_code) = 0;
+
+  /// Updates the state machine based on the handling of a cancelled transition
+  /**
+   * \param[in] success true if the transition cancel request was successfully handled
+   * \return true if the response was successfully sent to the state handler
+  */
+  virtual bool canceled(bool success) = 0;
```
Comment (alsora) [inline rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: the name of this method seems like we are just checking the state with no side-effects...
what about a different name such as `handle_canceled` ?

File: rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp
Comment (alsora) [inline rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: Moving this stuff to their own files is definitlely a good change, but is it required? can we do it as a separate PR?

File: rclcpp_lifecycle/src/change_state_handler_impl.cpp (hunk 2 above)
Comment (tgroechel) [reply rclcpp_lifecycle/src/change_state_handler_impl.cpp]: Will do, especially if preferred in `rclcpp`.

I personally prefer putting weak pointer lock [assignment in if statements to guarantee their lifespan](https://stackoverflow.com/questions/37679636/what-is-diffrence-between-lock-and-expired-weak-ptr-c) similar to empty brackets for a lock.

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp (hunk 3 above)
Comment (tgroechel) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: This is essentially what it was initially called (and I prefer from a readability standpoint). 

However, I decided to follow the same exact conventions of ROS 2 Action which has a `canceled(response)` [here](https://github.com/ros2/rclcpp/blob/rolling/rclcpp_action/include/rclcpp_action/server_goal_handle.hpp#L208) with a response (which is the analog to this function). Example of this being called [here](https://docs.ros.org/en/foxy/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.html#writing-the-action-server-code)

File: rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp
Comment (tgroechel) [reply rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: I would prefer to keep it within this PR. I think it retains readability/easier to see the direct transition to an action (when everything was in `lc_interface_impl`, it was very difficult for me to understand how things even worked in lifecycle).

It was a bit convoluted to do the change and I built the `StateManager` based on this. I can attempt an "unwind" if that makes the process easier to understand. The one thing I could easily put back into `lc_interface_imp` would be the `EntitiesManager` as it is somewhat unrelated (but don't see the benefit of doing so if I'm keeping the other MVC related changes).

**Comment by tgroechel** [comment]
- [ ] create unit tests
    - [x] register
    - [x] async w immediate return
    - [ ] async w deferred return
    - [ ] cancel
- [x] fix no throw error colcon test 
- [x] update demo code for `canceled`->handle_canceled`
- [x] unwind into outside branch for SoC (split into #4 & #5 )

File: rclcpp_lifecycle/src/change_state_handler_impl.cpp (hunk 2 above)
Comment (tgroechel) [reply rclcpp_lifecycle/src/change_state_handler_impl.cpp]: Done here 
https://github.com/tgroechel/rclcpp/pull/3/commits/33512a298f9d8ca48dfa1b975761d11a4feb852d

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp (hunk 3 above)
Comment (tgroechel) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: Resolving as `wont do` given naming scheme of actions (although I expect this will come back up in the `rclcpp` PR / agree handle canceled is better)

File: rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp
Comment (tgroechel) [reply rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: Resolving as `wont do` given the above

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp (hunk 1 above)
Comment (tgroechel) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: Done in https://github.com/tgroechel/rclcpp/pull/3/commits/b35d75524eaf207eb526bf6192dd73a09a9b414a (directly via github)

File: rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp (hunk 3 above)
Comment (tgroechel) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: Re-opening after thinking about it this weekend, `handeled_cancel` is much better

Comment (tgroechel) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp]: Done in
https://github.com/tgroechel/rclcpp/pull/3/commits/b442bff40b7f9831352aede655934975a0ad9da4

File: rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp
Comment (alsora) [reply rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: What about opening first a PR with the cleanup and then the PR with the new stuff?

Comment (tgroechel) [reply rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: I could do this. I agree that it would make maintenance/cherry picks for `rclcpp` easier if the SoC and async were separate.

For clarification, do you mean in sequence or in parallel for the PRs? I really would not want to do the former (the latter has additional (mild*) ongoing dual maintenance overhead).

The thing holding me back from doing this is that the ROS 2 community may just out right reject this PR's implementation or async transitions as a whole. At that point, I don't want to waste a lot of time on the implementation separation.

Comment (tgroechel) [reply rclcpp_lifecycle/src/lifecycle_node_state_services_manager.cpp:1]: I'll do this today, shouldn't be that bad actually if I destruct from this branch, added to TODOs https://github.com/tgroechel/rclcpp/pull/3#issuecomment-1585001041

**Comment by tgroechel** [comment]
see #4 & #5 (posting to `rclcpp`)
